### PR TITLE
ExtractFiles can overwrite destination folder

### LINKED ISSUE #11593: ExtractFiles should allow overwriting destination
## Required Information

Entering this information will route you directly to the right team and expedite traction.

**Question, Bug, or Feature?**  
bug

**Enter Task Name**: ExtractFiles  

## Environment
- Server : Azure Pipelines

- Agent : Private : Windows Server 2019
    
## Issue Description

Task fails if "Clean destination folder before extracting" is not checked and destination folder is not empty.

### Task logs
```
2019-10-19T12:13:02.9307087Z ##[section]Starting: Extract files 
2019-10-19T12:13:02.9446365Z ==============================================================================
2019-10-19T12:13:02.9446437Z Task         : Extract files
2019-10-19T12:13:02.9446466Z Description  : Extract a variety of archive and compression files such as .7z, .rar, .tar.gz, and .zip
2019-10-19T12:13:02.9446510Z Version      : 1.151.1
2019-10-19T12:13:02.9446551Z Author       : Microsoft Corporation
2019-10-19T12:13:02.9446577Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/extract-files
2019-10-19T12:13:02.9446604Z ==============================================================================
2019-10-19T12:13:03.2123419Z No path specified for search pattern: *.zip defaulting to: C:\azagent\A1\_work\r1\a
2019-10-19T12:13:03.2124863Z Searching for: *.zip under directory: C:\azagent\A1\_work\r1\a
2019-10-19T12:13:03.2200742Z Found: 1 files to extract:
2019-10-19T12:13:03.2200895Z C:\azagent\A1\_work\r1\a\Staging - CI\AssetToDeply\AssetToDeply.zip
2019-10-19T12:13:03.2207021Z Extracting file: C:\azagent\A1\_work\r1\a\Staging - CI\AssetToDeply\AssetToDeply.zip
2019-10-19T12:13:03.2219725Z [command]C:\azagent\A1\_work\_tasks\ExtractFiles_5e1e3830-fbfb-11e5-aab1-090c92bc4988\1.151.1\7zip\7z.exe x -oC:\Apps\AssetDest\dev C:\azagent\A1\_work\r1\a\Staging - CI\AssetToDeply\AssetToDeply.zip
2019-10-19T12:13:03.2435313Z 
2019-10-19T12:13:03.2435539Z 
2019-10-19T12:13:03.2435783Z ERROR: Can't allocate required memory!
2019-10-19T12:13:03.2435822Z 7-Zip [64] 16.00 : Copyright (c) 1999-2016 Igor Pavlov : 2016-05-10
2019-10-19T12:13:03.2435867Z 
2019-10-19T12:13:03.2435912Z Scanning the drive for archives:
2019-10-19T12:13:03.2435961Z 1 file, 35814372 bytes (35 MiB)
2019-10-19T12:13:03.2435994Z 
2019-10-19T12:13:03.2436042Z Extracting archive: C:\azagent\A1\_work\r1\a\Staging - CI\AssetToDeply\AssetToDeply.zip
2019-10-19T12:13:03.2436104Z --
2019-10-19T12:13:03.2436184Z Path = C:\azagent\A1\_work\r1\a\Staging - CI\AssetToDeply\AssetToDeply.zip
2019-10-19T12:13:03.2436216Z Type = zip
2019-10-19T12:13:03.2436292Z Physical Size = 35814372
2019-10-19T12:13:03.2436347Z 
2019-10-19T12:13:03.2436377Z 
2019-10-19T12:13:03.2436424Z Would you like to replace the existing file:
2019-10-19T12:13:03.2436475Z   Path:     C:\Apps\AssetDest\dev\Api.Abstractions.dll
2019-10-19T12:13:03.2436519Z   Size:     10240 bytes (10 KiB)
2019-10-19T12:13:03.2436548Z   Modified: 2019-10-19 10:16:40
2019-10-19T12:13:03.2436639Z with the file from archive:
2019-10-19T12:13:03.2436730Z   Path:     Api.Abstractions.dll
2019-10-19T12:13:03.2436761Z   Size:     10240 bytes (10 KiB)
2019-10-19T12:13:03.2436841Z   Modified: 2019-10-19 10:16:40
2019-10-19T12:13:03.2436915Z ? (Y)es / (N)o / (A)lways / (S)kip all / A(u)to rename all / (Q)uit? 
2019-10-19T12:13:03.2436981Z Archives with Errors: 1
2019-10-19T12:13:03.2437000Z 
2019-10-19T12:13:03.2438414Z code: 2
```

## Troubleshooting
Need to add "-aoa" switch (Overwrite All existing files without prompt.) to 7zip command.

## Related issues
- ExtractFiles failing in Release pipeline with ERROR: Can't allocate required memory! #11135
- ExtractFiles to support override files on the target path #5522

### PR DESCRIPTION
**Pull request by TFleury**
Fixes #11593

### CODE REVIEW

**Comment by stephenmichaelf** [comment]
Thanks for taking the time to add a fix! Please see the discussion [here](https://github.com/microsoft/azure-pipelines-tasks/pull/6760/files). Let's add an environment variable that can be used to disable the feature if need be.

**Comment by TFleury** [comment]
@stephenmichaelf Sorry, I didn't see the previous PR.
So, I added an environment variable to opt-out.

File: Tasks/ExtractFilesV1/extractfilestask.ts [hunk 1]
```diff
@@ -10,6 +10,9 @@ var cleanDestinationFolder: boolean = tl.getBoolInput('cleanDestinationFolder',
 var repoRoot: string = tl.getVariable('System.DefaultWorkingDirectory');
 tl.debug('repoRoot: ' + repoRoot);
 
+var overwriteDestinationVar: string = tl.getVariable('ExtractFile.OverwriteDestination');
```
Comment (stephenmichaelf) [inline Tasks/ExtractFilesV1/extractfilestask.ts]: Let's call this something like `EXTRACT_FILES_PREVENT_FILE_OVERWRITE`. The naming you are using is for another type of variable.

File: Tasks/ExtractFilesV1/extractfilestask.ts [hunk 2]
```diff
@@ -10,6 +10,9 @@ var cleanDestinationFolder: boolean = tl.getBoolInput('cleanDestinationFolder',
...
+var overwriteDestination: boolean = overwriteDestinationVar ? overwriteDestinationVar.toLowerCase() === 'true' : true;
```
Comment (stephenmichaelf) [inline Tasks/ExtractFilesV1/extractfilestask.ts]: Let's make this the opposite, let's default it to overwriting and only if the variable is set then we prevent it. This lets users opt-out if they want to but all other users get the overwriting.

**Comment by WilliamAntonRohm** [comment]
@stephenmichaelf -- Stephen, please review these changes, per [azure-devops-docs issue 8029](https://github.com/MicrosoftDocs/azure-devops-docs/issues/8029).  If you'd like, I could then re-create this PR for the current base branch.

**Comment by WilliamAntonRohm** [comment]
@stephenmichaelf -- Stephen, per the GitHub nag message above, please let me know if this PR needs to be recreated, or if your https://github.com/microsoft/azure-pipelines-tasks/pull/12902 sufficiently addresses https://github.com/microsoft/azure-pipelines-tasks/issues/11593, or if that original issue is no longer relevant.  Thanks!

**Comment by anatolybolshakov** [comment]
Hi @WilliamAntonRohm could you please check if it works for you now? We already added support for overwriting of existing files and changes have been rolled out.
@TFleury thanks for contribution here!

**Comment by anatolybolshakov** [comment]
Closing this at the moment since we have rolled out fix for original issue.

**Pull request closed without merging**